File: Data Structures & Algorithms/max-area-of-island/submission-1.py
from typing import List

class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        def dfs(i, j):
            if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] == 0:
                return 0
            
            ans = 1
            grid[i][j] = 0

            for direction in dirs:
                new_i = i + direction[0]
                new_j = j + direction[1]
                
                ans += dfs(new_i, new_j)
            return ans
        
        res = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    res = max(res, dfs(i, j))
        
        return res

File: Data Structures & Algorithms/burst-balloons/submission-1.cpp
class Solution
{
public:
    int maxCoins(vector<int> &nums)
    {
        vector<vector<int>> dp(nums.size() + 2, vector<int>(nums.size() + 2, -1));
        vector<int> new_nums = vector<int>(nums.size() + 2, 1);
        for (int i = 1; i <= nums.size(); i++)
        {
            new_nums[i] = nums[i - 1];
        }
        return dfs(1, nums.size(), new_nums, dp);
    }

    int dfs(int first, int last, vector<int> &nums, vector<vector<int>> &dp)
    {
        if (dp[first][last] != -1)
        {
            return dp[first][last];
        }
        int ans = 0;
        for (int i = first; i <= last; i++)
        {
            int temp = nums[i] * nums[first - 1] * nums[last + 1];
            if (i == first)
            {
                temp += dfs(i + 1, last, nums, dp);
            }
            else if (i == last)
            {
                temp += dfs(first, i - 1, nums, dp);
            }
            else
            {
                temp += dfs(first, i - 1, nums, dp) + dfs(i + 1, last, nums, dp);
            }
            ans = max(ans, temp);
        }
        dp[first][last] = ans;
        return ans;
    }
};